### ISSUE TITLE
What if there were no SQL?

### ISSUE THREAD

**Issue by sam-wheat** [dotnet/efcore#20028]
I wonder if you guys could answer a question that I have had for years.

What if there were a DBMS that could parse a LINQ query natively , without first translating it into SQL.  Would there be any benefit to EF users in terms of performance, reduced complexity, reliability, more LINQ flexibility, etc.

I ask this question because I have been using EF for many years and I see the bug reports that come through and it seems like you guys have to account for every nuance and possible variation in SQL.  I can only imagine how complex EF is by now.  Wouldn't it be easier to just skip SQL altogether?  Has there been any discussion or effort in this direction already?  Why would this be a bad idea?

Thank you.

**Comment by NetTecture** [comment]
Then you run small databases with limites scope? Because you keep taling about LINQ - and ignore the fact that in larger scnearios databases are often where multiple programs written in many languages meet. Object databases and OQL failed for that reason. Read up on the history of SQL to get a larger picture, starting with Cobb's original paper ("The relational Theorem", IIRC). Limiting a database to one language is not going to make you a friend in any larger scenario.

There is also another problem with your statement. "account for every nuance and possible variation in SQL." - the same is done by abandoning any database not being i.e. SQL Server. No more nuances. See, most nuances are between database products. And people want to use their database, so - you just told your customers to f*** off. That said, this sounds totally like what happens in the sql team now ,because they lobotomized their SQL generation and tell any customer not writing "hello world" to f*** off as their example code works. But - let's assume for a moment you would offload LINQ to the database. WHAT DATABASE? Not a single major database accepts LINQ queries at the moment. So, not only do you get a serious strategiy problem - you also have no database.

**Comment by ajcvickers** [comment]
@sam-wheat There is value in what you say. The impedance mismatch between the app language and the database would be reduced considerably. In some sense the in-memory database works this way.

That being said, I don't think it would remove the need to target existing databases, including all their different dialects of SQL, so ultimately I'm not sure that it would have enough value.

**Comment by sam-wheat** [comment]
I think The amount of effort that has gone into EF (and other ORMs) over the last decade gives enormous credibility to the argument "There is huge value in writing your business logic in strongly typed code".  There is an impedance mismatch as you say, or, if you rephrase it a little, a very large opportunity for improvement.  

I see EF as an attempt to bridge the impedance mismatch.  It has turned out to be a long bridge and not a very desirable one.  Not because of EF or the team but because of the nature of the solution itself.  It's a band-aid.  Its a great band-aid, IMO the best band-aid that is available. But it is still a band-aid (please understand that I am not in any way complaining about EF).

What if we greatly reduce the mismatch and bring code and DBMS closer together? What would we do differently?  The solution to the problem might not be compatible with existing databases but so what?  It would be a huge step forward and a compelling reason to use .net.  NoSQL, Hadoop, Hive, Spark, MapReduce, etc. seem to be doing fine and these are platforms that have pushed the envelope.

**Comment by ajcvickers** [comment]
@sam-wheat I don't disagree that there is room to be innovative and push the envelope here. :-) I'm just saying it won't remove the need to help people access existing databases.

**Comment by NetTecture** [comment]
@sam-wheat @ajcvickers That discussion would be even a little funny if this whole discussion was not done and closed 15 years ago with Java and Object Databases. Been there, see them fail and nothing has changed. The mismatch is systemic in the different requirements and not in the language. Look up the old discussions from 25 years ago. There is nothing innovative in retrying failed experiments unless you can come up with arguments that are different than the one from back then.

Google for "OQL" and generally "OBject databases".

**Comment by sam-wheat** [comment]
@ajcvickers A valid point.

**Comment by sam-wheat** [comment]
@NetTecture  You don't understand my question.  I am not proposing an end to relational databases .  I am saying what if instead of doing this:

    linq -> SQL -> parsed query the DBMS understands

we did this

    linq -> parsed query the DBMS understands


Perhaps internally the DBMS generates SQL or some intermediate.  Who cares.


> Limiting a database to one language is not going to make you a friend in any larger scenario.

	string sql = "SELECT * FROM CUSTOMERS";
	string linq = "from c in db.customers select c";

**Comment by roji** [comment]
Some additional notes to complement what @ajcvickers wrote above:

The EF Core model is already about translating directly to a "parsed query the DBMS understands", without passing through LINQ. For example, if we wanted to support database X which understands some NoSQL language called Y, EF Core wouldn't be used to first generate SQL and then translate that into Y - the provider model would allow us to translate directly from LINQ to Y. In fact, EF Core was explicitly designed from the ground up to separate SQL and relational concepts from the core, explicitly in order to support NoSQL scenarios. In nuget terms, all core functionality is in the [Microsoft.EntityFrameworkCore](https://www.nuget.org/packages/Microsoft.EntityFrameworkCore) nuget package, while relational concepts (including all the SQL-related complexity) are in [Microsoft.EntityFrameworkCore.Relational](https://www.nuget.org/packages/Microsoft.EntityFrameworkCore.Relational/). The latter package wouldn't even be referenced by a NoSQL provider.

On the other hand, I'm not aware of any database which accepts .NET LINQ queries (e.g. as strings) natively, as you seem to be suggesting - that would make the database quite .NET-specific (how would one interact with it from, say, Java)? Some databases do come with a .NET package which includes a LINQ provider (to improve the experience of their .NET users only), but the database itself invariably has some other, lower-level query language which the LINQ provider translates to. For these cases, an EF Core provider could (at least theoretically) be written to use the LINQ interface, though there are various reasons why translating to the lower-level language may be a better idea.

To summarize, I'm not sure there's anything really to do here, and EF Core should already be architected in a way which allows interacting with NoSQL databases without the complexity of SQL impacting it in any way.

**Comment by sam-wheat** [comment]
>  I'm not sure there's anything really to do here, 

There is nothing to do.  This is a question intended to get the right people thinking about the possibilities.  I have posted this question in the past on forums where I have been yelled at and mocked.  I have long wanted to post it here where I could get thoughtful responses from the people most qualified to give them (thanks!).  


> which allows interacting with NoSQL

To be clear, I used the words "no SQL" in the question but my intent is to ask something much more general like "Is there a faster way to get a query which uses linq syntax into a format a DBMS can actually execute and bypass translation to any human readable query of any other language or syntax".


What got me thinking about this question again is the work @SteveSandersonMS is doing on Blazor.  @ajcvickers mentioned the impedance mismatch - there has been a similar impedance mismatch between web servers and web browsers.  We basically had to code our model in c# and again in JS to make them work.  Now that we have Blazor we have extended dotnet right into the browser.  How cool is that!  Now, why can't we do extend dotnet into the DBMS?

> On the other hand, I'm not aware of any database which accepts .NET LINQ queries (e.g. as strings) natively, as you seem to be suggesting - 

Exactly - this is the core of my question.  Why is there no database that can parse my linq query?


> that would make the database quite .NET-specific

Yes it would. However dotnet core is an awesome framework.  I want to use it everywhere!  

> how would one interact with it from, say, Java)?

Well, as it stands now I have to use an X database and a shim (EF) if I want to write linq.  Why can't I have a linq database and let the other guy use a shim if he wants to write X? 

Having said all this I have no idea how any of it can be done.  I am posting here because you guys are far more qualified to answer these questions.  
Speculation and wishful thinking:

* Code up my model in a nuget (I do that now already) then add my nuget to the DBMS to create my schema/migrations.

* Code up my business logic in a nuget and add that nuget to the DBMS.  Then I could write server side linq like .where(x => MyFunction(x)).  

* Write my business logic/tests in c# from some equivalent of SSMS where I could write linq in a query window.

* etc....

> I think the amount of effort that has gone into EF (and other ORMs) over the last decade gives enormous credibility to the argument "There is huge value in writing your business logic in strongly typed code".

Sorry for repeating myself but this is the driver behind all this.  EF is a game changer.  dotnet core is a game changer.  While EF is not new I feel like its mission (writing business logic in strongly typed code) has never been elevated, prioritized, and appreciated like it should be.  

EF as it exists today is a band-aid.  It should be a platform.  I really feel like there is work that EF can do to eliminate the impedance mismatch (as Blazor has done) rather than bridge it.  I hope there are visionaries at Microsoft who can make it happen.

**Comment by roji** [comment]
Some additional comments:

* First off, it's worth mentioning that what various databases support is up to them - we have no influence over that (we just do EF on this repo).
* One thing you may be missing here, is that LINQ isn't about writing a query in a string and sending that somewhere for execution (as you seem to be suggesting in [this comment](https://github.com/dotnet/efcore/issues/20028#issuecomment-589997218)). The whole point is for your query to be verified statically - and that requires your .NET program to have a model of your data, in code, at compile time. This is exactly what EF Core provides - a code model and the ability to express LINQ queries over that. If you sent LINQ text queries directly to a database, you would basically be back to writing SQL in many ways, just with a different syntax.
* LINQ is a way to express queries, but EF Core does more than just querying; it handles change detection, updates and many other things. Even if a database exposed a native LINQ interface for querying, those other non-querying aspects could be good reasons for using EF Core.
* Finally, what specific advantage do you think you'd gain by communicating LINQ directly to the database, as opposed to going through EF Core?

I'm going to close this issue now as there's nothing immediately actionable on the EF Core side, but feel free to ask further questions.

**Comment by sam-wheat** [comment]
> One thing you may be missing here, is that LINQ isn't about writing a query in a string

I am not missing it at all.  It is the reason for this entire post. The question I was addressing was how might a Java developer call a linq server.  

EF, starting with linq-to-sql has been under development for ~12 years.  Look at where we are today. We have a translator that is brittle and fragile.  I am not complaining about EF! It is doing what it supposed to do - I am saying that the constraints on its design are self-defeating.  

The mission and purpose of EF - writing business logic in statically typed code - is profoundly important.

The goal of writing a dbms that is accessible to the world with a shim to leverage dotnet is wrong.

The goal of writing a dbms that is leverages dotnet is the correct one.  If a Java developer wants to use it let him use a shim.  In truth these are probably not mutually exclusive objectives.  I just want to make a point about the importance of leveraging dotnet.


>Finally, what specific advantage do you think you'd gain by communicating LINQ directly to the database, as opposed to going through EF Core?

How awesome would it be to write linq and use c# functions as first-class citizens on the server?

Can you even imagine how cool it would be to write linq in c# classes and deploy them to a server as a nuget?  

Wouldn't it be great to eliminate the burden of translating to sql (or NoSQL or X) and just let the dbms run queries in a manner that is most direct and efficient?


It is sad to think that none of this will ever happen.  Microsoft is going to do the same thing with EF that they did with the web.  The last web developers left Microsoft ten years ago.  They had to jump through too many hoops to use dotnet to write web apps.  Now Microsoft has Blazor - a wonderful product - and there is only a handful of people left who care.

The same will happen with EF.  It is a work-around that while well-crafted takes a wrong approach.  In ten years someone at Microsoft will say "Hey look at this dbms that runs dotnet natively".  There will be no one there to listen. @SteveSandersonMS  can you hear me?
